### PR TITLE
clarify that unchecked indexing is UB even if the reference is never used

### PR DESCRIPTION
**Pull request by RalfJung**

### CODE REVIEW

File: src/libcore/slice/mod.rs [hunk 1]
```diff
@@ -292,8 +292,10 @@ impl<T> [T] {
     /// Returns a reference to an element or subslice, without doing bounds
     /// checking.
     ///
-    /// This is generally not recommended, use with caution! For a safe
-    /// alternative see [`get`].
+    /// This is generally not recommended, use with caution!
+    /// Calling this method with an out-of-bounds index is UB even if the resulting
```
Comment (Centril) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// Calling this method with an out-of-bounds index is *[undefined behavior]*
```

File: src/libcore/slice/mod.rs [hunk 2]
```diff
@@ -292,8 +292,10 @@ impl<T> [T] {
...
+    /// reference is not used.
```
Comment (Centril) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// even if the resulting reference is not used.
```

File: src/libcore/slice/mod.rs [hunk 3]
```diff
@@ -292,8 +292,10 @@ impl<T> [T] {
...
+    /// For a safe alternative see [`get`].
     ///
     /// [`get`]: #method.get
```
Comment (Centril) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// [`get`]: #method.get
    /// [undefined behavior]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
```

File: src/libcore/slice/mod.rs [hunk 4]
```diff
@@ -317,8 +319,10 @@ impl<T> [T] {
     /// Returns a mutable reference to an element or subslice, without doing
     /// bounds checking.
     ///
-    /// This is generally not recommended, use with caution! For a safe
-    /// alternative see [`get_mut`].
+    /// This is generally not recommended, use with caution!
+    /// Calling this method with an out-of-bounds index is UB even if the resulting
```
Comment (Centril) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// Calling this method with an out-of-bounds index is *[undefined behavior]*
```

File: src/libcore/slice/mod.rs [hunk 5]
```diff
@@ -317,8 +319,10 @@ impl<T> [T] {
...
+    /// reference is not used.
```
Comment (Centril) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// even if the resulting reference is not used.
```

File: src/libcore/slice/mod.rs [hunk 6]
```diff
@@ -317,8 +319,10 @@ impl<T> [T] {
...
+    /// For a safe alternative see [`get_mut`].
     ///
     /// [`get_mut`]: #method.get_mut
```
Comment (Centril) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// [`get_mut`]: #method.get_mut
    /// [undefined behavior]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
```

File: src/libcore/slice/mod.rs [hunk 7]
```diff
@@ -2629,11 +2633,15 @@ pub trait SliceIndex<T: ?Sized>: private_slice_index::Sealed {
 
     /// Returns a shared reference to the output at this location, without
     /// performing any bounds checking.
+    /// Calling this method with an out-of-bounds index is UB even if the resulting
```
Comment (Centril) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// Calling this method with an out-of-bounds index is *[undefined behavior]*
```

File: src/libcore/slice/mod.rs [hunk 8]
```diff
@@ -2629,11 +2633,15 @@ pub trait SliceIndex<T: ?Sized>: private_slice_index::Sealed {
...
+    /// reference is not used.
```
Comment (Centril) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// even if the resulting reference is not used.
    /// [undefined behavior]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
```

File: src/libcore/slice/mod.rs [hunk 9]
```diff
@@ -2629,11 +2633,15 @@ pub trait SliceIndex<T: ?Sized>: private_slice_index::Sealed {
...
     #[unstable(feature = "slice_index_methods", issue = "0")]
     unsafe fn get_unchecked(self, slice: &T) -> &Self::Output;
 
     /// Returns a mutable reference to the output at this location, without
     /// performing any bounds checking.
+    /// Calling this method with an out-of-bounds index is UB even if the resulting
+    /// reference is not used.
```
Comment (Centril) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// even if the resulting reference is not used.
    /// [undefined behavior]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
```

File: src/libcore/slice/mod.rs [hunk 10]
```diff
@@ -2629,11 +2633,15 @@ pub trait SliceIndex<T: ?Sized>: private_slice_index::Sealed {
...
     /// Returns a shared reference to the output at this location, without
     /// performing any bounds checking.
+    /// Calling this method with an out-of-bounds index is UB even if the resulting
+    /// reference is not used.
     #[unstable(feature = "slice_index_methods", issue = "0")]
     unsafe fn get_unchecked(self, slice: &T) -> &Self::Output;
 
     /// Returns a mutable reference to the output at this location, without
     /// performing any bounds checking.
+    /// Calling this method with an out-of-bounds index is UB even if the resulting
```
Comment (Centril) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// Calling this method with an out-of-bounds index is *[undefined behavior]*
```

**Review by Centril — COMMENTED** [review]
Thanks! r=me rollup with suggestions applied.

File: src/libcore/slice/mod.rs (hunk 3 above)
Comment (RalfJung) [reply src/libcore/slice/mod.rs]: I took the liberty to update the reference link to be relative (so it stays on the version of the rust docs one was looking at).

**Comment by RalfJung** [comment]
@bors r-

investigating

**Comment by RalfJung** [comment]
Okay I give up, we do indeed have to use an absolute link.

**Pull request merged**